main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1069 @@
+            c.unifyTypeM(tvar, lambdaBodyType, loc)
+            c.unifyTypeM(evar, Type.mkUnion(lambdaBodyEff :: eff :: effs, loc), loc)
+            val resTpe = tvar
+            val resEff = evar
+            (resTpe, resEff)
+        }
+
+      case Expr.Lambda(fparam, exp, loc) =>
+        c.unifyTypeM(fparam.sym.tvar, fparam.tpe, loc)
+        val (tpe, eff) = visitExp(exp)
+        val resTpe = Type.mkArrowWithEffect(fparam.tpe, eff, tpe, loc)
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case KindedAst.Expr.Unary(sop, exp, tvar, loc) => sop match {
+        case SemanticOp.BoolOp.Not =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Bool, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Bool, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case SemanticOp.Float32Op.Neg =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Float32, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Float32, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case SemanticOp.Float64Op.Neg =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Float64, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Float64, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case SemanticOp.Int8Op.Neg | SemanticOp.Int8Op.Not =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Int8, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Int8, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case SemanticOp.Int16Op.Neg | SemanticOp.Int16Op.Not =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Int16, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Int16, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case SemanticOp.Int32Op.Neg | SemanticOp.Int32Op.Not =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Int32, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Int32, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case SemanticOp.Int64Op.Neg | SemanticOp.Int64Op.Not =>
+          val (tpe, eff) = visitExp(exp)
+          c.expectTypeM(expected = Type.Int64, actual = tpe, exp.loc)
+          c.unifyTypeM(Type.Int64, tvar, exp.loc)
+          val resTpe = tvar
+          val resEff = eff
+          (resTpe, resEff)
+
+        case _ => throw InternalCompilerException(s"Unexpected unary operator: '$sop'.", loc)
+      }
+
+      case KindedAst.Expr.Binary(sop, exp1, exp2, tvar, loc) => sop match {
+
+        case SemanticOp.BoolOp.And | SemanticOp.BoolOp.Or =>
```

Comment:
These could also all share a helper function, then itd be easier to read IMO, the reader just have to verify the hardcoded types of each operation, not all the copied code.

Reply:
listed